synapse/event_auth.py

```diff
@@ -1054,9 +1054,14 @@ def _verify_third_party_invite(
     """
     if "third_party_invite" not in event.content:
         return False
-    if "signed" not in event.content["third_party_invite"]:
+    third_party_invite = event.content["third_party_invite"]
+    if not isinstance(third_party_invite, collections.abc.Mapping):
+        return False
+    if "signed" not in third_party_invite:
+        return False
+    signed = third_party_invite["signed"]
+    if not isinstance(signed, collections.abc.Mapping):
         return False
-    signed = event.content["third_party_invite"]["signed"]
     for key in {"mxid", "token"}:
```

Comment:
Should we create a dedicated type with these required fields? Or some magic interim Python mapping type where we can require fields.

`signed = Signed(mxid, signatures, token)`

Reply:
I don't think we pull this out anywhere else, so it likely wouldn't heavily simplify the code.